### PR TITLE
Adding namespace option to microDDS client

### PR DESCRIPTION
**Pull request by dsix-ls2n**
This pr allows adding a **namespace** to the microDDS client.
The participant will be named /**namespace**/px4_xrce_participant while the topics will appear as **/namespace**/fmu/...
This allows having multiple instances of the PX4 Autopilot on the same network.
The namespace can be set as an optional argument of the microdds client. 
The default namespace is empty and does not affect the current topic names.

### CODE REVIEW

**Comment by TSC21** [comment]
@dsix-ls2n can you run `make format` to fix the format, please? Thanks!

File: msg/templates/uorb_microcdr/dds_topics.h.em [hunk 1]
```diff
@@ -48,11 +48,11 @@ struct SendTopicsSubs {
 
 	uint32_t num_payload_sent{};
 
-	bool init(uxrSession* session_, uxrStreamId stream_id, uxrObjectId participant_id);
+	bool init(uxrSession* session_, uxrStreamId stream_id, uxrObjectId participant_id, const std::string &client_namespace);
```
Comment (bkueng) [inline msg/templates/uorb_microcdr/dds_topics.h.em:51]: `std::string` isn't available on embedded.
```suggestion
	bool init(uxrSession* session_, uxrStreamId stream_id, uxrObjectId participant_id, const const char* client_namespace);
```

**Review by bkueng — COMMENTED** [review]
Thanks.
Only partially related, but this would be a good opportunity to switch to the [binary interface](https://micro-xrce-dds.docs.eprosima.com/en/latest/client_api.html#entities-bin-api), if you're interested.

File: msg/templates/uorb_microcdr/dds_topics.h.em (hunk 1 above)
Comment (dsix-ls2n) [reply msg/templates/uorb_microcdr/dds_topics.h.em:51]: Sure, my bad, I will solve that. cont char* is really not convenient for concatenation but I didn't test embedded afterward.
The binary interface should be nicer too, I will have a look.

**Comment by dsix-ls2n** [comment]
> Thanks. Only partially related, but this would be a good opportunity to switch to the [binary interface](https://micro-xrce-dds.docs.eprosima.com/en/latest/client_api.html#entities-bin-api), if you're interested.

Hello, I am looking into this binary interface. It's clear for most of it. However, I didn't find out how to set a specific transport for the participant as in the following xml part with the binary API. If anyone knows, it will be a pleasure to finish this.
"<dds>"
"<profiles>"
"<transport_descriptors>"
"<transport_descriptor>"
"<transport_id>udp_localhost</transport_id>"
"<type>UDPv4</type>"
"<interfaceWhiteList><address>127.0.0.1</address></interfaceWhiteList>"
"</transport_descriptor>"
"</transport_descriptors>"
"</profiles>"
"<participant>"
"<rtps>"
            "<name>" + _client_namespace + "px4_xrce_participant</name>"
"<useBuiltinTransports>false</useBuiltinTransports>"
"<userTransports><transport_id>udp_localhost</transport_id></userTransports>"
"</rtps>"
"</participant>"
"</dds>"
:
"<dds>"
"<participant>"
"<rtps>"
            "<name>" + _client_namespace + "px4_xrce_participant</name>"
"</rtps>"
"</participant>"
"</dds>" ;

**Comment by dsix-ls2n** [comment]
I pushed a fix. I couldn't find a solution to use the binary interface for all cases. So I let the xml interface for participant creation. Normally, it should compile on board now. I couldn't make a full test with the board but unfortunately, I have surgery tomorrow so I let it as it is for now.

File: msg/templates/uorb_microcdr/dds_topics.h.em [hunk 2]
```diff
@@ -63,33 +63,27 @@ topic_pascal = topic.replace("_", " ").title().replace(" ", "")
...
-				"<topic>"
-				"<name>rt/fmu/out/@(topic_pascal)</name>"
-				"<dataType>px4_msgs::msg::dds_::@(topic_pascal)_</dataType>"
-				"</topic>"
-				"</dds>";
-		uint16_t topic_req = uxr_buffer_create_topic_xml(session, stream_id, topic_id, participant_id, topic_xml,
-						UXR_REPLACE);
+		const char* topic_postfix = "fmu/out/@(topic_pascal)";
+		uint16_t topic_name_size = strlen(topic_postfix) + strlen(client_namespace) + 4;
+        char* topic_name = new char [topic_name_size];
```
Comment (bkueng) [inline msg/templates/uorb_microcdr/dds_topics.h.em:68]: This isn't deleted. Not so relevant here, but you could use a fixed-size array:
```
	char* topic[128]
	int n = snprintf(topic, sizeof(topic), "rt/%sfmu/out/%s", _client_namespace, "@(topic_pascal)");

	if (n >= sizeof(topic)) {
		PX4_ERR("topic path too long");
```

File: src/modules/microdds_client/microdds_client.cpp [hunk 3]
```diff
@@ -52,9 +52,20 @@
 using namespace time_literals;
 
 MicroddsClient::MicroddsClient(Transport transport, const char *device, int baudrate, const char *host,
-			       const char *port, bool localhost_only)
+			       const char *port, bool localhost_only, const char *client_namespace)
 	: _localhost_only(localhost_only)
 {
+	if (client_namespace == nullptr) {
+		_client_namespace = new char [1];
```
Comment (bkueng) [inline src/modules/microdds_client/microdds_client.cpp:59]: This needs to be deleted in the destructor

**Review by bkueng — COMMENTED** [review]
> I couldn't find a solution to use the binary interface for all cases.

Let's leave it at that then.

> I couldn't make a full test with the board but unfortunately, I have surgery tomorrow so I let it as it is for now.

I hope all goes well!

**Comment by dagar** [comment]
Rolled into https://github.com/PX4/PX4-Autopilot/pull/20227.

**Pull request closed without merging**